invoices/update_invoice_test.go

```diff
@@ -0,0 +1,690 @@
+			input: InvoiceHTLC{
+				Amt:           5000,
+				MppTotalAmt:   5000,
+				AcceptHeight:  100,
+				AcceptTime:    testNow,
+				ResolveTime:   testAlreadyNow,
+				Expiry:        40,
+				State:         HtlcStateCanceled,
+				CustomRecords: make(record.CustomSet),
+				AMP: &InvoiceHtlcAMPData{
+					Record:   *ampRecord,
+					Hash:     hash,
+					Preimage: &preimage,
+				},
+			},
+			invState: ContractAccepted,
+			setID:    &setID,
+			output: InvoiceHTLC{
+				Amt:           5000,
+				MppTotalAmt:   5000,
+				AcceptHeight:  100,
+				AcceptTime:    testNow,
+				ResolveTime:   testAlreadyNow,
+				Expiry:        40,
+				State:         HtlcStateCanceled,
+				CustomRecords: make(record.CustomSet),
+				AMP: &InvoiceHtlcAMPData{
+					Record:   *ampRecord,
+					Hash:     hash,
+					Preimage: &preimage,
+				},
+			},
+			expErr: nil,
+		},
+		{
+			name: "cancel invoice htlc already canceled",
+			input: InvoiceHTLC{
+				Amt:           5000,
+				MppTotalAmt:   5000,
+				AcceptHeight:  100,
+				AcceptTime:    testNow,
+				ResolveTime:   testAlreadyNow,
+				Expiry:        40,
+				State:         HtlcStateCanceled,
+				CustomRecords: make(record.CustomSet),
+				AMP: &InvoiceHtlcAMPData{
+					Record:   *ampRecord,
+					Hash:     hash,
+					Preimage: &preimage,
+				},
+			},
+			invState: ContractCanceled,
+			setID:    &setID,
+			output: InvoiceHTLC{
+				Amt:           5000,
+				MppTotalAmt:   5000,
+				AcceptHeight:  100,
+				AcceptTime:    testNow,
+				ResolveTime:   testAlreadyNow,
+				Expiry:        40,
+				State:         HtlcStateCanceled,
+				CustomRecords: make(record.CustomSet),
+				AMP: &InvoiceHtlcAMPData{
+					Record:   *ampRecord,
+					Hash:     hash,
+					Preimage: &preimage,
+				},
+			},
+			expErr: nil,
+		},
+		{
+			name: "settle invoice htlc already canceled",
+			input: InvoiceHTLC{
+				Amt:           5000,
+				MppTotalAmt:   5000,
+				AcceptHeight:  100,
+				AcceptTime:    testNow,
+				ResolveTime:   testAlreadyNow,
+				Expiry:        40,
+				State:         HtlcStateCanceled,
+				CustomRecords: make(record.CustomSet),
+				AMP: &InvoiceHtlcAMPData{
+					Record:   *ampRecord,
+					Hash:     hash,
+					Preimage: &preimage,
+				},
+			},
+			invState: ContractSettled,
+			setID:    &setID,
+			output: InvoiceHTLC{
+				Amt:           5000,
+				MppTotalAmt:   5000,
+				AcceptHeight:  100,
+				AcceptTime:    testNow,
+				ResolveTime:   testAlreadyNow,
+				Expiry:        40,
+				State:         HtlcStateCanceled,
+				CustomRecords: make(record.CustomSet),
+				AMP: &InvoiceHtlcAMPData{
+					Record:   *ampRecord,
+					Hash:     hash,
+					Preimage: &preimage,
+				},
+			},
+			expErr: nil,
+		},
+	}
+
+	for _, test := range tests {
+		test := test
+		t.Run(test.name, func(t *testing.T) {
+			testUpdateHTLC(t, test)
+		})
+	}
+}
+
+func testUpdateHTLC(t *testing.T, test updateHTLCTest) {
+	htlc := test.input.Copy()
+	stateChanged, state, err := getUpdatedHtlcState(
```

Comment:
looks like only `getUpdatedHtlcState` is tested and we should also test for other newly added methods/functions.

Reply:
Right, next commit moves all the invoice tests which provide the necessary coverage.

The idea with the refactor really was to do equivalent updates to the code such that we can keep the existing test coverage while introducing the `InvoiceUpdater` interface.